### DISCUSSION TITLE
Topocentric to geographic projection

### DISCUSSION THREAD

**Discussion by WangYanyan2020** [pyproj4/pyproj#1348] [Q&A]
I'm trying to convert a topocentric coordinate to a geographic one. 
Referring to other posts, I found that the inverse of this transformation (i.e., geographic to topocentric) can be made by making a Transformer from a pipeline as the following: 

pipe_trans = pyproj.Transformer.from_pipeline("+proj=pipeline +step +proj=cart +ellps=WGS84 +step +proj=topocentric +ellps=WGS84 +lon_0=136 +lat_0=33 +h_0=0)

I expected that the other way around, as

pipe_trans_inv = pyproj.Transformer.from_pipeline("+proj=pipeline +step +proj=topocentric +ellps=WGS84 +lon_0=136 +lat_0=33 +h_0=0 +step +proj=cart +ellps=WGS84")

works for my target transformation, but actually does not work and gives an error as follows. 

ProjError: Invalid projection +proj=pipeline +step +proj=topocentric +ellps=WGS84 +lon_0=136 +lat_0=33 +h_0=0 +step +proj=cart +ellps=WGS84.: (Internal Proj Error: proj_create: Error 1025 (Invalid PROJ string syntax): pipeline: Pipeline: Mismatched units between step 1 and 2)

How should I fix this error? I don't persist on using the pipeline if I can make the transformation using pyproj.

Thank you!

**Comment by djhoese** [comment]
I don't know enough about PROJ pipelines to give a good answer, but does this operation work with PROJ's command line tools? It seems to be an error with the definition and not pyproj itself so I'm wondering if PROJ's documentation can shed more light on the problem...at least until someone who has more experience with this type of definition can answer.

**Reply by WangYanyan2020** [reply]
Thank you for your reply! 
I will also take a look at PROJ's documentation more carefully.

**Comment by snowman2** [comment]
The PROJ mailing list is a good resource for a question like this: https://lists.osgeo.org/mailman/listinfo/proj/

**Comment by snowman2** [comment]
Related: https://github.com/pyproj4/pyproj/discussions/863

**Reply by WangYanyan2020** [reply]
Thank you for your notice! 
This is exactly the most related issue to this topic.

**Comment by xiemeilong** [comment]
Has your issue been resolved? I'm experiencing the same problem.

**Comment by jjimenezshaw** [comment]
Converting from a topocentric to EPSG:4326 (geographic lat-lon in degrees) is
```
+proj=pipeline
  +step +inv +proj=topocentric +lat_0=55 +lon_0=5 +h_0=0 +ellps=WGS84
  +step +inv +proj=cart +ellps=WGS84
  +step +proj=unitconvert +xy_in=rad +xy_out=deg
  +step +proj=axisswap +order=2,1
```

The conversion from EPSG:4326 (geographic) to topocentric is
```
+proj=pipeline
  +step +proj=axisswap +order=2,1
  +step +proj=unitconvert +xy_in=deg +z_in=m +xy_out=rad +z_out=m
  +step +proj=cart +ellps=WGS84
  +step +proj=topocentric +lat_0=55 +lon_0=5 +h_0=0 +ellps=WGS84
```

**Comment by xiemeilong** [comment]
Thank you very much, it works.
